docs/RFCS/20210527_changefeed_http_sink.md

```diff
@@ -0,0 +1,202 @@
+(`experimental-http`) uses the cloud storage sink implementation (see 
+[sink_cloudstorage.go](../../pkg/ccl/changefeedccl/sink_cloudstorage.go)). Many 
+Enterprise customers have expressed an interest in a webhook-type HTTP sink as
+opposed to the current implementation, which should be reserved for traditional
+cloud storage destinations (i.e. GCS/S3). 
+
+# Technical design
+
+## High-level Design
+
+The HTTP sink implements the existing 
+[`Sink`](../../pkg/ccl/changefeedccl/sink.go) interface with the same Sink 
+functions `EmitRow()`, `Flush()`, etc. and should offer the same guarantees 
+i.e. ordering, at-least-once delivery as other sinks. However, this sink 
+differs slightly in that it iss webhook-based and therefore stateless, so the 
+endpoints don't have to be health-checked and maintained throughout the 
+lifespan of the changefeed. So `Dial()` and `Close()`in the `Sink` interface 
+are no-ops in this scenario, at least for the sink itself (channel still needs 
+to be closed).
+
+The majority of the http connection logic should be wrapped in an 
+[`httputil.Client`](../../pkg/util/httputil/client.go) object (wrapper around Go's 
+[`http.Client`](https://golang.org/src/net/http/client.go#L57)) which will
+handle most of the TLS and other config setup mentioned below. The HTTP sink
+struct might look something like this:
+
+```
+type httpSink struct {
+  u sinkURL
+  client *httputil.Client
+	mu struct {
+		syncutil.Mutex
+		inflight int64
+		flushCh chan struct{}
+	}
+}
+```
+
+## Concurrency
+
+Each `POST` to the HTTP endpoint is executed asynchronously through calls to 
+`EmitRow()` and `EmitResolvedTimestamp()`. We use a channel here, `flushCh`,
+which we will send the http response object once it is received, or the error
+(if one occurs). We use a channel as opposed to something like a `WaitGroup` to
+receive a response (which may be needed for ack functionality later on) or an 
+error.
+
+```
+func (s *httpSink) EmitRow (...) error {
+  ...
+  res, err := s.client.Post(...)
+  s.mu.Lock()
+  s.mu.inflight++
+  s.mu.Unlock()
+  go func() {
+    s.mu.flushCh <- struct {
+      res *http.Response
+      err error
+    } {
+      res: res,
+      err: err,
+    }
+  }
+  ...
+  return err
+}
+```
+
+On the `Flush()` side, the the `mu` field can act as a semaphore, and allow
+reading from the channel until the sink has been flushed.
+
+```
+func (s *httpSink) Flush (...) error {
+  ...
+  s.mu.Lock()
+  defer s.mu.Unlock()
+  for s.mu.inflight > 0 {
+    select {
+      case ctx.Done():
+        return ctx.Err()
+      case flushed := <- s.mu.flushCh:
+        s.mu.inflight--
+        // can add additional ack logic here with flushed.res if needed
+        log.Infof("Sink returned response: %v", flushed.res)
+  }
+  ...
+}
+```
+
+## Configurability
+
+The HTTP sink should be configurable through a few different options and sink
+URL query params, a few of which are already supported by other sinks:
+
```

Comment:
I think we should spend a lot more time on this section. Perhaps it's more of a question for @amruss .  I feel we have a lot of space to explore here.
For example:
  * Should we support GET? POST? Or both?
  * If we support GET, which URL parameters do we send (key/value? key/value/mtime? each column a separate parameter)?  How configurable should that be? HOW to configure that is another question (from UI perspective).
  * If we're using POST, should we allow sending json/avro payload (probably)
  * If we allow POST, should we allow some sort of batching behavior (like kafka)? If so, how do you specify/configure it?
  * How configurable should the sink URI be (i.e. include modification time as a url parameter)?
 
Each additional option should be evaluated w/ an eye toward usability -- how do we express this in changefeed statement? Is it ambiguous, etc.

Reply:
it was my assumption that since we're using webhooks, it would be POST-only (i.e. send key/value as a JSON payload), but I'll let @amruss clarify it more.

In terms of query params in the sink URI, the plan is to let the user pass whatever params are needed, so long as they don't conflict with the ones already in place (mentioned in the 'open questions' section). we can add more specific query params as needed - but there may be conflicts, @stevendanna  mentioned this before

Reply:
> it was my assumption that since we're using webhooks, it would be POST-only (i.e. send key/value as a JSON payload), but I'll let @amruss clarify it more.

This is what makes the most sense to me - would like to get a few more users to weigh in though

Reply:
Regarding query parameters: I think we should consider not using query parameters for our options here for all the reasons you mention below in your Unanswered questions point on this subject.